### ISSUE TITLE
How to continue training the yolov8l-seg model but only on the detection task

### ISSUE THREAD

**Issue by IGlace** [ultralytics/ultralytics#8720]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I searched a lot for a model that can provide me with both results of detection and segmentation at the same time, and finally yolov8l-seg was able to provide me with i want.

The segmentation part of the model is perfect and i have no problems with it, but the detection task is not good enough for me, after deep analysis of the architecture in the onnx version of the model, i managed to find excatly what are the layers that are responsible on prediction of bounding boxes and classes.

So my question is :

1. How to freeze all the layers unrelated to the bounding box and class prediction task, like the backbone layers and the first head layers and the segmentations layers.
2. How to continue training the model on the detection task only, i saw the normal code provided in the documentation to train the model whole in general, but how i'm supposed to force it to only train that specific part.
3. What kind of annotations i should give him, the dataset i'm trying to give me him is a dataset of 45000 images with their annotations as bounding box coordinates and their class, like the same way how you would train on a normal yolov8l model for detection only, unlike what the yolov8l-seg require a polygonne coordinates for segmentation.

Here is my attempt first to freeze all the layers unrelated to detection task but i don't know how to continue from here, let me know please if i'm going in the right direction here : 
```
from ultralytics import YOLO
model = YOLO('yolov8l-seg.pt')
# Freeze
freeze_size = 22
freeze = [f'model.{x}.' for x in range(freeze_size)] + ['model.22.proto.','model.22.cv4.'] # layers to freeze
for k, v in model.named_parameters():
    v.requires_grad = True  # train all layers
    if any(x in k for x in freeze):
        v.requires_grad = False
```

**Comment by Burhan-Q** [comment]
@IGlace glad you're appreciating the segmentation model results. A few import notes:

> [!IMPORTANT]
> Once training completes, either from no model improvement or reaching the final epoch, **it's not possible to "continue" training** that will provide you with additive performance.

1. The `resume` argument is valid for any model task, but is _only_ valid for training that have been interrupted. So if the model finishes it's training cycle naturally, your model is complete and additional training will not be helpful. If you were to lose power part way thru training, you could continue training by loading the model weights `last.pt` and adding `resume=True` when calling the `train()` method.

1. You also can't train detection and segmentation separately, and my understanding is (and this _could be_ wrong) that the bounding boxes for a segmentation model are directly linked to the segmentation masks. 

1. Segmentation models should get _only_ segmentation annotations. Mixing annotation types will default to keeping bounding boxes only and training results will likely be poor.

**Comment by IGlace** [comment]
Hello, 

So i think you misunderstood my question a little bit, my objective is continue the training of the detection part only in the segmentation model, using a dataset for object detection.

The steps i followed for this is  : 

1. i loaded the segmentation model yolov8l-seg, and transformed the weights of it's layers from 0 to 22 to the yolov8l model for object detection.
2. Using the following code i trained the new yolov8l with changed weights on my own dataset for object dataset but by freezing the first 22 layers from 0 to 21 leaving only the layer 22 for object detection to continue training it : 

```
from ultralytics import YOLO
import os

os.environ["CUDA_VISIBLE_DEVICES"] = "0,1,2"

# Load a model
model = YOLO('updated_yolov8l.pt')  # load a pretrained model (recommended for training)

# Train the model with 3 GPUs
print("Start Training : ")
print("########################")
model.train(data='custom_dataset.yaml', epochs=100, imgsz=640, device=[0,1,2], batch=15, freeze=22, project="detection_freezed")
```

3. After my model finished training i took the weights of the layer 22 for detection, and transformed them back to the segmentation model in the layer 22 for detection.

My objective from this is fine tunning the detection part in the segmentation model, to work more pefrectly on my dataset, but i don't have data to train to the segmentation part so i want to keep it as it is, from my understanding from layer 0 to 21 it is the same architecture between the segmentation model and detection model which why it was the part i freezed to not affect the segmentation part in the segmentation model, yet i didn't get the results i want, please tell me what was wrong with my method, from my analysis the segmentation model is simply an extension of the detection model with additional layers running in parallel with detection layer where both of them get their input from the layers 15,18,21 .

If you think my method is wrong, please propose for me another method i can do this.

**Comment by Burhan-Q** [comment]
@IGlace you are correct that all layers for YOLOv8 object-detection and segmentation models up to layer index 22 are the same. I now believe I understand what you intend to accomplish here, but I'm a bit confused as to why exactly; not that I _need_ to.

> [!CAUTION]
> I'll start with saying, this is **not a supported** process and there's **no assurance** that it will work at all or as expected. This is also for anyone seeing this issue in the future. The official advice is that it would be best (and probably faster) to re-train the model. 

With that out of the way, I'll try to help out but I'll share that I'm not too familiar with the process, so I may have to get someone else to chime in. Let me start by asking how you're transferring your weights? Please share a code snippet as it will help me understand better.

In addition, I don't know if your data is similar to the COCO dataset, but you could also experiment with transferring weights from the COCO pretrained model to your custom segmentation model to see if that might help at all.

**Comment by IGlace** [comment]
@Burhan-Q To answer your first question why i need this, simply because i like the segmentation part of the model and it's really good, but the detection part is not good enough for me specifcly for the tasks i want to carry on after getting that detection results.

Using a dataset i prepared i managed to create a detection model giving me the results i want, it's just about merging the detection i want with the segmentation i want in one model, that's my objective here, sadly if i had a segmentation annotations for my dataset i would have done that and retrained the model fully, but since i don't i'm trying to find an alternative.

Now here is the code you asked for how i transfer the weights from the 'yolov8l-seg' model to 'yolov8l' model : 

```
import torch
from ultralytics import YOLO

# Instantiate the YOLOv8 model
source_model = YOLO('yolov8l-seg.pt')
target_model = YOLO('yolov8l.pt')

model_a_parameters = dict(source_model.named_parameters())

# Dictionary to hold the differences
differences = {}
count = 0
# Check for differences in module presence
for name, paramter in target_model.named_parameters():
    if name not in model_a_parameters:
        differences[name] = "Present in yolov8l but not in yolov8l-seg."
    else:
        if not torch.equal(paramter.data, model_a_parameters[name].data):
            differences[name] = "Parameter values differ."
            paramter.data = model_a_parameters[name].data
            count += 1

print(count)
print(differences)

target_model.save("updated_yolov8l.pt")
```

Here is a comparaison of before and after updating the weights : 

```
############ Before Change ###################
303
{'model.model.0.conv.weight': 'Parameter values differ.', 'model.model.0.bn.weight': 'Parameter values differ.', 'model.model.0.bn.bias': 'Parameter values differ.', 
'model.model.1.conv.weight': 'Parameter values differ.', 'model.model.1.bn.weight': 'Parameter values differ.', 'model.model.1.bn.bias': 'Parameter values differ.', 'model.model.2.cv1.conv.weight': 'Parameter values differ.', 'model.model.2.cv1.bn.weight': 'Parameter values differ.', 'model.model.2.cv1.bn.bias': 'Parameter values 
differ.', 'model.model.2.cv2.conv.weight': 'Parameter values differ.', 'model.model.2.cv2.bn.weight': 'Parameter values differ.', 'model.model.2.cv2.bn.bias': 'Parameter values differ.', 'model.model.2.m.0.cv1.conv.weight': 'Parameter values differ.', 'model.model.2.m.0.cv1.bn.weight': 'Parameter values differ.', 'model.model.2.m.0.cv1.bn.bias': 'Parameter values differ.', 'model.model.2.m.0.cv2.conv.weight': 'Parameter values differ.', 'model.model.2.m.0.cv2.bn.weight': 'Parameter values differ.', 'model.model.2.m.0.cv2.bn.bias': 'Parameter values differ.', 'model.model.2.m.1.cv1.conv.weight': 'Parameter values differ.', 'model.model.2.m.1.cv1.bn.weight': 'Parameter values differ.', 'model.model.2.m.1.cv1.bn.bias': 'Parameter values differ.', 'model.model.2.m.1.cv2.conv.weight': 'Parameter values differ.', 'model.model.2.m.1.cv2.bn.weight': 'Parameter values differ.', 'model.model.2.m.1.cv2.bn.bias': 'Parameter values differ.', 'model.model.2.m.2.cv1.conv.weight': 'Parameter values differ.', 'model.model.2.m.2.cv1.bn.weight': 'Parameter values differ.', 'model.model.2.m.2.cv1.bn.bias': 'Parameter values differ.', 'model.model.2.m.2.cv2.conv.weight': 'Parameter values differ.', 'model.model.2.m.2.cv2.bn.weight': 'Parameter values differ.', 'model.model.2.m.2.cv2.bn.bias': 'Parameter values differ.', 'model.model.3.conv.weight': 'Parameter values differ.', 'model.model.3.bn.weight': 'Parameter values differ.', 'model.model.3.bn.bias': 'Parameter values differ.', 'model.model.4.cv1.conv.weight': 'Parameter values differ.', 'model.model.4.cv1.bn.weight': 'Parameter values differ.', 'model.model.4.cv1.bn.bias': 'Parameter values differ.', 'model.model.4.cv2.conv.weight': 'Parameter values differ.', 'model.model.4.cv2.bn.weight': 'Parameter values differ.', 'model.model.4.cv2.bn.bias': 'Parameter values differ.', 'model.model.4.m.0.cv1.conv.weight': 'Parameter values differ.', 'model.model.4.m.0.cv1.bn.weight': 'Parameter values differ.', 'model.model.4.m.0.cv1.bn.bias': 'Parameter values differ.', 'model.model.4.m.0.cv2.conv.weight': 'Parameter values differ.', 'model.model.4.m.0.cv2.bn.weight': 'Parameter values differ.', 'model.model.4.m.0.cv2.bn.bias': 'Parameter values differ.', 'model.model.4.m.1.cv1.conv.weight': 'Parameter values differ.', 'model.model.4.m.1.cv1.bn.weight': 'Parameter values differ.', 'model.model.4.m.1.cv1.bn.bias': 'Parameter values differ.', 'model.model.4.m.1.cv2.conv.weight': 'Parameter values differ.', 'model.model.4.m.1.cv2.bn.weight': 'Parameter values differ.', 'model.model.4.m.1.cv2.bn.bias': 'Parameter values differ.', 'model.model.4.m.2.cv1.conv.weight': 'Parameter values differ.', 'model.model.4.m.2.cv1.bn.weight': 'Parameter values differ.', 'model.model.4.m.2.cv1.bn.bias': 'Parameter values differ.', 'model.model.4.m.2.cv2.conv.weight': 
'Parameter values differ.', 'model.model.4.m.2.cv2.bn.weight': 'Parameter values differ.', 'model.model.4.m.2.cv2.bn.bias': 'Parameter values differ.', 'model.model.4.m.3.cv1.conv.weight': 'Parameter values differ.', 'model.model.4.m.3.cv1.bn.weight': 'Parameter values differ.', 'model.model.4.m.3.cv1.bn.bias': 'Parameter values differ.', 'model.model.4.m.3.cv2.conv.weight': 'Parameter values differ.', 'model.model.4.m.3.cv2.bn.weight': 'Parameter values differ.', 'model.model.4.m.3.cv2.bn.bias': 'Parameter values differ.', 'model.model.4.m.4.cv1.conv.weight': 'Parameter values differ.', 'model.model.4.m.4.cv1.bn.weight': 'Parameter values differ.', 'model.model.4.m.4.cv1.bn.bias': 'Parameter values differ.', 'model.model.4.m.4.cv2.conv.weight': 'Parameter values differ.', 'model.model.4.m.4.cv2.bn.weight': 'Parameter 
values differ.', 'model.model.4.m.4.cv2.bn.bias': 'Parameter values differ.', 'model.model.4.m.5.cv1.conv.weight': 'Parameter values differ.', 'model.model.4.m.5.cv1.bn.weight': 'Parameter values differ.', 'model.model.4.m.5.cv1.bn.bias': 'Parameter values differ.', 'model.model.4.m.5.cv2.conv.weight': 'Parameter values differ.', 
'model.model.4.m.5.cv2.bn.weight': 'Parameter values differ.', 'model.model.4.m.5.cv2.bn.bias': 'Parameter values differ.', 'model.model.5.conv.weight': 'Parameter values differ.', 'model.model.5.bn.weight': 'Parameter values differ.', 'model.model.5.bn.bias': 'Parameter values differ.', 'model.model.6.cv1.conv.weight': 'Parameter values differ.', 'model.model.6.cv1.bn.weight': 'Parameter values differ.', 'model.model.6.cv1.bn.bias': 'Parameter values differ.', 'model.model.6.cv2.conv.weight': 'Parameter values differ.', 'model.model.6.cv2.bn.weight': 'Parameter values differ.', 'model.model.6.cv2.bn.bias': 'Parameter values differ.', 'model.model.6.m.0.cv1.conv.weight': 'Parameter values differ.', 'model.model.6.m.0.cv1.bn.weight': 'Parameter values differ.', 'model.model.6.m.0.cv1.bn.bias': 'Parameter values differ.', 'model.model.6.m.0.cv2.conv.weight': 'Parameter values differ.', 'model.model.6.m.0.cv2.bn.weight': 'Parameter values differ.', 'model.model.6.m.0.cv2.bn.bias': 'Parameter values differ.', 'model.model.6.m.1.cv1.conv.weight': 'Parameter values differ.', 'model.model.6.m.1.cv1.bn.weight': 'Parameter values differ.', 'model.model.6.m.1.cv1.bn.bias': 'Parameter values differ.', 'model.model.6.m.1.cv2.conv.weight': 'Parameter values differ.', 'model.model.6.m.1.cv2.bn.weight': 'Parameter values 
differ.', 'model.model.6.m.1.cv2.bn.bias': 'Parameter values differ.', 'model.model.6.m.2.cv1.conv.weight': 'Parameter values differ.', 'model.model.6.m.2.cv1.bn.weight': 'Parameter values differ.', 'model.model.6.m.2.cv1.bn.bias': 'Parameter values differ.', 'model.model.6.m.2.cv2.conv.weight': 'Parameter values differ.', 'model.model.6.m.2.cv2.bn.weight': 'Parameter values differ.', 'model.model.6.m.2.cv2.bn.bias': 'Parameter values differ.', 'model.model.6.m.3.cv1.conv.weight': 'Parameter values differ.', 'model.model.6.m.3.cv1.bn.weight': 'Parameter values differ.', 'model.model.6.m.3.cv1.bn.bias': 'Parameter values differ.', 'model.model.6.m.3.cv2.conv.weight': 'Parameter values differ.', 'model.model.6.m.3.cv2.bn.weight': 'Parameter values differ.', 'model.model.6.m.3.cv2.bn.bias': 'Parameter values differ.', 'model.model.6.m.4.cv1.conv.weight': 'Parameter values differ.', 'model.model.6.m.4.cv1.bn.weight': 'Parameter values differ.', 'model.model.6.m.4.cv1.bn.bias': 'Parameter values differ.', 'model.model.6.m.4.cv2.conv.weight': 'Parameter values differ.', 'model.model.6.m.4.cv2.bn.weight': 'Parameter values differ.', 'model.model.6.m.4.cv2.bn.bias': 'Parameter values differ.', 'model.model.6.m.5.cv1.conv.weight': 'Parameter values differ.', 'model.model.6.m.5.cv1.bn.weight': 'Parameter values differ.', 'model.model.6.m.5.cv1.bn.bias': 'Parameter values differ.', 'model.model.6.m.5.cv2.conv.weight': 'Parameter values differ.', 'model.model.6.m.5.cv2.bn.weight': 
'Parameter values differ.', 'model.model.6.m.5.cv2.bn.bias': 'Parameter values differ.', 'model.model.7.conv.weight': 'Parameter values differ.', 'model.model.7.bn.weight': 'Parameter values differ.', 'model.model.7.bn.bias': 'Parameter values differ.', 'model.model.8.cv1.conv.weight': 'Parameter values differ.', 'model.model.8.cv1.bn.weight': 'Parameter values differ.', 'model.model.8.cv1.bn.bias': 'Parameter values differ.', 'model.model.8.cv2.conv.weight': 'Parameter values differ.', 'model.model.8.cv2.bn.weight': 'Parameter values differ.', 'model.model.8.cv2.bn.bias': 'Parameter values differ.', 'model.model.8.m.0.cv1.conv.weight': 'Parameter values differ.', 'model.model.8.m.0.cv1.bn.weight': 'Parameter values differ.', 'model.model.8.m.0.cv1.bn.bias': 'Parameter values differ.', 'model.model.8.m.0.cv2.conv.weight': 'Parameter values differ.', 'model.model.8.m.0.cv2.bn.weight': 'Parameter values differ.', 'model.model.8.m.0.cv2.bn.bias': 'Parameter values differ.', 'model.model.8.m.1.cv1.conv.weight': 'Parameter values differ.', 'model.model.8.m.1.cv1.bn.weight': 'Parameter values differ.', 'model.model.8.m.1.cv1.bn.bias': 'Parameter values differ.', 'model.model.8.m.1.cv2.conv.weight': 'Parameter values differ.', 'model.model.8.m.1.cv2.bn.weight': 'Parameter values differ.', 'model.model.8.m.1.cv2.bn.bias': 'Parameter values differ.', 'model.model.8.m.2.cv1.conv.weight': 'Parameter values differ.', 'model.model.8.m.2.cv1.bn.weight': 'Parameter values differ.', 'model.model.8.m.2.cv1.bn.bias': 'Parameter values differ.', 'model.model.8.m.2.cv2.conv.weight': 'Parameter values differ.', 'model.model.8.m.2.cv2.bn.weight': 'Parameter values differ.', 'model.model.8.m.2.cv2.bn.bias': 'Parameter values differ.', 'model.model.9.cv1.conv.weight': 'Parameter values differ.', 'model.model.9.cv1.bn.weight': 'Parameter values differ.', 'model.model.9.cv1.bn.bias': 'Parameter values differ.', 'model.model.9.cv2.conv.weight': 'Parameter values differ.', 'model.model.9.cv2.bn.weight': 'Parameter values differ.', 'model.model.9.cv2.bn.bias': 'Parameter values differ.', 'model.model.12.cv1.conv.weight': 'Parameter values differ.', 
'model.model.12.cv1.bn.weight': 'Parameter values differ.', 'model.model.12.cv1.bn.bias': 'Parameter values differ.', 'model.model.12.cv2.conv.weight': 'Parameter values differ.', 'model.model.12.cv2.bn.weight': 'Parameter values differ.', 'model.model.12.cv2.bn.bias': 'Parameter values differ.', 'model.model.12.m.0.cv1.conv.weight': 'Parameter values differ.', 'model.model.12.m.0.cv1.bn.weight': 'Parameter values differ.', 'model.model.12.m.0.cv1.bn.bias': 'Parameter values differ.', 'model.model.12.m.0.cv2.conv.weight': 'Parameter values differ.', 'model.model.12.m.0.cv2.bn.weight': 'Parameter values differ.', 'model.model.12.m.0.cv2.bn.bias': 'Parameter values differ.', 'model.model.12.m.1.cv1.conv.weight': 'Parameter values differ.', 'model.model.12.m.1.cv1.bn.weight': 'Parameter values differ.', 'model.model.12.m.1.cv1.bn.bias': 'Parameter values differ.', 'model.model.12.m.1.cv2.conv.weight': 'Parameter values differ.', 'model.model.12.m.1.cv2.bn.weight': 'Parameter values differ.', 'model.model.12.m.1.cv2.bn.bias': 'Parameter values differ.', 'model.model.12.m.2.cv1.conv.weight': 'Parameter values differ.', 'model.model.12.m.2.cv1.bn.weight': 'Parameter values differ.', 'model.model.12.m.2.cv1.bn.bias': 'Parameter values differ.', 'model.model.12.m.2.cv2.conv.weight': 'Parameter values differ.', 'model.model.12.m.2.cv2.bn.weight': 'Parameter values differ.', 'model.model.12.m.2.cv2.bn.bias': 'Parameter values differ.', 'model.model.15.cv1.conv.weight': 'Parameter values differ.', 'model.model.15.cv1.bn.weight': 'Parameter values differ.', 'model.model.15.cv1.bn.bias': 'Parameter values differ.', 'model.model.15.cv2.conv.weight': 'Parameter values differ.', 'model.model.15.cv2.bn.weight': 'Parameter values differ.', 'model.model.15.cv2.bn.bias': 'Parameter values differ.', 'model.model.15.m.0.cv1.conv.weight': 'Parameter values differ.', 'model.model.15.m.0.cv1.bn.weight': 'Parameter values differ.', 'model.model.15.m.0.cv1.bn.bias': 'Parameter values 
differ.', 'model.model.15.m.0.cv2.conv.weight': 'Parameter values differ.', 'model.model.15.m.0.cv2.bn.weight': 'Parameter values differ.', 'model.model.15.m.0.cv2.bn.bias': 'Parameter values differ.', 'model.model.15.m.1.cv1.conv.weight': 'Parameter values differ.', 'model.model.15.m.1.cv1.bn.weight': 'Parameter values differ.', 
'model.model.15.m.1.cv1.bn.bias': 'Parameter values differ.', 'model.model.15.m.1.cv2.conv.weight': 'Parameter values differ.', 'model.model.15.m.1.cv2.bn.weight': 'Parameter values differ.', 'model.model.15.m.1.cv2.bn.bias': 'Parameter values differ.', 'model.model.15.m.2.cv1.conv.weight': 'Parameter values differ.', 'model.model.15.m.2.cv1.bn.weight': 'Parameter values differ.', 'model.model.15.m.2.cv1.bn.bias': 'Parameter values differ.', 'model.model.15.m.2.cv2.conv.weight': 'Parameter values differ.', 'model.model.15.m.2.cv2.bn.weight': 'Parameter values differ.', 'model.model.15.m.2.cv2.bn.bias': 'Parameter values differ.', 'model.model.16.conv.weight': 'Parameter values differ.', 'model.model.16.bn.weight': 'Parameter values differ.', 'model.model.16.bn.bias': 'Parameter values differ.', 'model.model.18.cv1.conv.weight': 'Parameter values differ.', 'model.model.18.cv1.bn.weight': 'Parameter values differ.', 'model.model.18.cv1.bn.bias': 'Parameter values differ.', 'model.model.18.cv2.conv.weight': 'Parameter values differ.', 'model.model.18.cv2.bn.weight': 'Parameter values differ.', 'model.model.18.cv2.bn.bias': 'Parameter values differ.', 'model.model.18.m.0.cv1.conv.weight': 'Parameter values differ.', 'model.model.18.m.0.cv1.bn.weight': 'Parameter values differ.', 'model.model.18.m.0.cv1.bn.bias': 'Parameter values differ.', 'model.model.18.m.0.cv2.conv.weight': 'Parameter values differ.', 'model.model.18.m.0.cv2.bn.weight': 'Parameter values differ.', 'model.model.18.m.0.cv2.bn.bias': 'Parameter values differ.', 'model.model.18.m.1.cv1.conv.weight': 'Parameter values differ.', 'model.model.18.m.1.cv1.bn.weight': 'Parameter values differ.', 'model.model.18.m.1.cv1.bn.bias': 'Parameter values differ.', 'model.model.18.m.1.cv2.conv.weight': 'Parameter values differ.', 'model.model.18.m.1.cv2.bn.weight': 'Parameter values differ.', 'model.model.18.m.1.cv2.bn.bias': 'Parameter values differ.', 'model.model.18.m.2.cv1.conv.weight': 'Parameter values differ.', 'model.model.18.m.2.cv1.bn.weight': 'Parameter values differ.', 'model.model.18.m.2.cv1.bn.bias': 'Parameter values differ.', 'model.model.18.m.2.cv2.conv.weight': 'Parameter values differ.', 'model.model.18.m.2.cv2.bn.weight': 'Parameter values differ.', 'model.model.18.m.2.cv2.bn.bias': 'Parameter values differ.', 'model.model.19.conv.weight': 'Parameter values differ.', 'model.model.19.bn.weight': 'Parameter values differ.', 'model.model.19.bn.bias': 'Parameter values differ.', 'model.model.21.cv1.conv.weight': 'Parameter values differ.', 'model.model.21.cv1.bn.weight': 'Parameter values differ.', 'model.model.21.cv1.bn.bias': 'Parameter values 
differ.', 'model.model.21.cv2.conv.weight': 'Parameter values differ.', 'model.model.21.cv2.bn.weight': 'Parameter values differ.', 'model.model.21.cv2.bn.bias': 'Parameter values differ.', 'model.model.21.m.0.cv1.conv.weight': 'Parameter values differ.', 'model.model.21.m.0.cv1.bn.weight': 'Parameter values differ.', 'model.model.21.m.0.cv1.bn.bias': 'Parameter values differ.', 'model.model.21.m.0.cv2.conv.weight': 'Parameter values differ.', 'model.model.21.m.0.cv2.bn.weight': 'Parameter values differ.', 'model.model.21.m.0.cv2.bn.bias': 'Parameter values differ.', 'model.model.21.m.1.cv1.conv.weight': 'Parameter values differ.', 'model.model.21.m.1.cv1.bn.weight': 'Parameter values differ.', 'model.model.21.m.1.cv1.bn.bias': 'Parameter values differ.', 'model.model.21.m.1.cv2.conv.weight': 'Parameter values differ.', 'model.model.21.m.1.cv2.bn.weight': 'Parameter values differ.', 'model.model.21.m.1.cv2.bn.bias': 'Parameter values differ.', 'model.model.21.m.2.cv1.conv.weight': 
'Parameter values differ.', 'model.model.21.m.2.cv1.bn.weight': 'Parameter values differ.', 'model.model.21.m.2.cv1.bn.bias': 'Parameter values differ.', 'model.model.21.m.2.cv2.conv.weight': 'Parameter values differ.', 'model.model.21.m.2.cv2.bn.weight': 'Parameter values differ.', 'model.model.21.m.2.cv2.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.0.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.0.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.0.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.0.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.0.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.0.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.0.2.weight': 'Parameter values differ.', 'model.model.22.cv2.0.2.bias': 'Parameter values differ.', 'model.model.22.cv2.1.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.1.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.1.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.1.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.1.1.bn.weight': 'Parameter values 
differ.', 'model.model.22.cv2.1.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.1.2.weight': 'Parameter values differ.', 'model.model.22.cv2.1.2.bias': 'Parameter values differ.', 'model.model.22.cv2.2.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.2.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.2.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.2.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.2.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.2.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.2.2.weight': 'Parameter values differ.', 'model.model.22.cv2.2.2.bias': 'Parameter values differ.', 'model.model.22.cv3.0.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.0.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.0.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.0.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.0.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.0.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.0.2.weight': 'Parameter values differ.', 'model.model.22.cv3.0.2.bias': 'Parameter values differ.', 'model.model.22.cv3.1.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.1.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.1.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.1.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.1.1.bn.weight': 
'Parameter values differ.', 'model.model.22.cv3.1.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.1.2.weight': 'Parameter values differ.', 'model.model.22.cv3.1.2.bias': 'Parameter values differ.', 'model.model.22.cv3.2.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.2.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.2.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.2.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.2.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.2.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.2.2.weight': 'Parameter values differ.', 'model.model.22.cv3.2.2.bias': 'Parameter values differ.'}
############ After Change ###################
0
{}
```

With this i create a model named 'updated_yolov8l.pt' with the weights for it's 23 layers from the segmentation model, of course since the model.22.cv4 and model.22.proto layers that exist in the segmentation model doesn't exist in the detection model it simply will not carry on.

Now as from what i copy my weights, as i told you above i wanna keep the segmentations results as good as they are, so i don't wanna touch the layers from 0 to 21 that the segmentation layers cv4 and proto take as input from, i want to keep them the same i want to only find tune the weights of the layers model.22.cv2 and model.22.cv3 for detection. And yes as an initial test to figure out if my method can work i trained the detection model with freezed layers from 0 to 21 with the coco dateset for object detection and did give me good results when doing test directly on the detection model, yet when i transfer the weights of the layers model.22.cv2 and model.22.cv3 for detection back to my segmentation model the results change in a bad way altough the weights of the layers from 0 to 21 are still the same in both of them.

As evidence when compared the difference of weights between the original segmentation model and my new updated segmentation model here is the resutls : 

```
********************** Check Difference Between Updated Segmentation Model and Original Segmentation model ****************** 
Number of differences : 48 {'model.model.22.cv2.0.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.0.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.0.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.0.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.0.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.0.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.0.2.weight': 'Parameter values differ.', 'model.model.22.cv2.0.2.bias': 'Parameter values differ.', 'model.model.22.cv2.1.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.1.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.1.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.1.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.1.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.1.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.1.2.weight': 'Parameter values differ.', 'model.model.22.cv2.1.2.bias': 'Parameter values differ.', 'model.model.22.cv2.2.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.2.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.2.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.2.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv2.2.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv2.2.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv2.2.2.weight': 'Parameter values differ.', 'model.model.22.cv2.2.2.bias': 'Parameter values differ.', 'model.model.22.cv3.0.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.0.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.0.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.0.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.0.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.0.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.0.2.weight': 'Parameter values differ.', 'model.model.22.cv3.0.2.bias': 'Parameter values differ.', 'model.model.22.cv3.1.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.1.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.1.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.1.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.1.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.1.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.1.2.weight': 'Parameter values differ.', 'model.model.22.cv3.1.2.bias': 'Parameter values differ.', 'model.model.22.cv3.2.0.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.2.0.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.2.0.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.2.1.conv.weight': 'Parameter values differ.', 'model.model.22.cv3.2.1.bn.weight': 'Parameter values differ.', 'model.model.22.cv3.2.1.bn.bias': 'Parameter values differ.', 'model.model.22.cv3.2.2.weight': 'Parameter values differ.', 'model.model.22.cv3.2.2.bias': 'Parameter values differ.'}
```

Yet the results are not good at all i also checked how the segment head work in the source code, it doesn't do anything special it just take the same input it would have sent to detection block if it was alone to the detection block again in segmentation block : 

![image](https://github.com/ultralytics/ultralytics/assets/60987472/ecd6d1b2-51ca-4631-9fd0-7eceb07461ce)

And finally thanks for your help with this, and hope i made everything clear.

**Comment by Burhan-Q** [comment]
@IGlace so to be clear, you're looking to _only_ update `model.22.cv2` and `model.22.cv3`? If not, the code I have below should be changed to filter the keys you wish to update specifically. One aspect that I noticed as a difference when comparing the model weights from Segment and Detection (pretrained) is that the pretrained `segment` model contains `model.22.cv2.0.2.bias` that is not in the `detect` model (pretrained `yolov8n-seg.pt` on left and pretrained `yolov8n.pt` on the right).

![image](https://github.com/ultralytics/ultralytics/assets/62214284/e1232fd1-5a5e-441c-aeb6-471fea417294)

Additionally the segmentation layers (after `model.22.dfl.conv.weight`) are not present in the detection model either, but I think you're aware of this. It's my suspicion that the missing component from the detection model may cause issues when attempting to update the weights in the segmentation model, but I don't honestly know.

I believe this code will accomplish what you're looking for, however the result for me was not very good. You might have to experiment with which layers are updated or not to find which works best for you; I did try using `"model.22" in k` instead, which was _slightly_ better, but still did not perform well. If I am to venture a guess as to why the poor performance occurs, I think it could have to do with differences in the `v8SegmentationLoss` and `v8DetectionLoss` but to be honest, this is far beyond my understanding at the moment.

```py
from collections import OrderedDict
from ultralytics import YOLO

detect = YOLO("yolov8n.pt") # pretrained detection
segment = YOLO("yolov8n-seg.pt") # pretrained segmentation

new_weights = OrderedDict()
keys = segment.state_dict().keys()
for k,w in detect.state_dict().items():
    if k in keys and "model.22.cv2" in k or "model.22.cv3" in k: # only update specific layers
        new_weights.update({k:w})

_ = segment.load_state_dict(new_weights, strict=False)
```

Overall since this isn't a supported process, I don't know that I can offer much more help for you. However if you are able to figure out a good method to accomplish this that is repeatable, I'm sure that others in the community would be curious to hear what you did that was successful. Personally I think that using `state_dict()` and `load_state_dict()` to be a quick/easy way to try transferring the weights as it will transfer both parameters and buffers, and using the `strict=False` argument will ignore the incompatible keys and even return a `NamedTuple` of the `missing` or `unexpected` keys.

**Comment by IGlace** [comment]
@Burhan-Q Thank you very much that method to transfer weights with state_dict worked like a charm, and transfering the weights from the segmentation model for the layers 0-21 and layer 22 for detection to a detection model, produced the same detection results as the segmentation model.

But you see the problem i stumbled with is now since i have the detection model with weights from the segmentation model and i want to freeze the layers from 0-21 and only train the layer 22 for detection, to do that i used the freeze parameters in the train function to freeze the first 22 layers in the model using this code : 

```
from ultralytics import YOLO

# Load a model
model = YOLO('updated_yolov8l.pt')  # load a pretrained model

# Train the model with 3 GPUs
print("Start Training : ")
print("########################")
model.train(data='custom_dataset.yaml', epochs=300, imgsz=640, batch=25, freeze=22, project="custom_detection")
```

Now you see the training finished perfectly with good results, and good precision, but after that i wanted to make a comparaison between their layers from 0-21 from the model before training and after training, to make sure the layers were freezed correctly, 

To do so, i save the inputs passed from layer to layer into model in the middile of the inference for each model using pickle by adding this simple change in the ultralytics code : 

![image](https://github.com/ultralytics/ultralytics/assets/60987472/11687ead-de7f-4ec2-aa0c-d9214d0d9fee)

After applying this to the two models seperatly, and comparing the inputs passing from each model, they are completly different which mean something changed even if the layers were freezed, and just to let you know i'm sure this method work, because i used this to compare between the inputs in the layers between the original segmentation model and the new detection model that i transfered the weights to, and showed me that the inputs going in are the same in each layer, yet now is different starting from the second layer since the first layer input is just the image anyway.

The final test i did is comparing between the stat_dict of the two detection models, and see what is different turn out the layers paramters were correctly freezed, but what's changed my results is something else, here is the log i printed for the paramters that hold different value in both state_dict of the two models except of course the final layer 22 because it's gonna be different of course since that's why i trained the model, that's why i hope you can explain to me why those paramters change even after i freezed those layers, they messing with everything i wanna do and i can't continue if the outputs of the final layer 21 is different after training i need it to be the same to follow my plan, please intsruct me how to freeze those parameters and what are they anyway.

```
model.model.0.bn.running_mean
model.model.0.bn.running_var
model.model.1.bn.running_mean
model.model.1.bn.running_var
model.model.2.cv1.bn.running_mean
model.model.2.cv1.bn.running_var
model.model.2.cv2.bn.running_mean
model.model.2.cv2.bn.running_var
model.model.2.m.0.cv1.bn.running_mean
model.model.2.m.0.cv1.bn.running_var
model.model.2.m.0.cv2.bn.running_mean
model.model.2.m.0.cv2.bn.running_var
model.model.2.m.1.cv1.bn.running_mean
model.model.2.m.1.cv1.bn.running_var
model.model.2.m.1.cv2.bn.running_mean
model.model.2.m.1.cv2.bn.running_var
model.model.2.m.2.cv1.bn.running_mean
model.model.2.m.2.cv1.bn.running_var
model.model.2.m.2.cv2.bn.running_mean
model.model.2.m.2.cv2.bn.running_var
model.model.3.bn.running_mean
model.model.3.bn.running_var
model.model.4.cv1.bn.running_mean
model.model.4.cv1.bn.running_var
model.model.4.cv2.bn.running_mean
model.model.4.cv2.bn.running_var
model.model.4.m.0.cv1.bn.running_mean
model.model.4.m.0.cv1.bn.running_var
model.model.4.m.0.cv2.bn.running_mean
model.model.4.m.0.cv2.bn.running_var
model.model.4.m.1.cv1.bn.running_mean
model.model.4.m.1.cv1.bn.running_var
model.model.4.m.1.cv2.bn.running_mean
model.model.4.m.1.cv2.bn.running_var
model.model.4.m.2.cv1.bn.running_mean
model.model.4.m.2.cv1.bn.running_var
model.model.4.m.2.cv2.bn.running_mean
model.model.4.m.2.cv2.bn.running_var
model.model.4.m.3.cv1.bn.running_mean
model.model.4.m.3.cv1.bn.running_var
model.model.4.m.3.cv2.bn.running_mean
model.model.4.m.3.cv2.bn.running_var
model.model.4.m.4.cv1.bn.running_mean
model.model.4.m.4.cv1.bn.running_var
model.model.4.m.4.cv2.bn.running_mean
model.model.4.m.4.cv2.bn.running_var
model.model.4.m.5.cv1.bn.running_mean
model.model.4.m.5.cv1.bn.running_var
model.model.4.m.5.cv2.bn.running_mean
model.model.4.m.5.cv2.bn.running_var
model.model.5.bn.running_mean
model.model.5.bn.running_var
model.model.6.cv1.bn.running_mean
model.model.6.cv1.bn.running_var
model.model.6.cv2.bn.running_mean
model.model.6.cv2.bn.running_var
model.model.6.m.0.cv1.bn.running_mean
model.model.6.m.0.cv1.bn.running_var
model.model.6.m.0.cv2.bn.running_mean
model.model.6.m.0.cv2.bn.running_var
model.model.6.m.1.cv1.bn.running_mean
model.model.6.m.1.cv1.bn.running_var
model.model.6.m.1.cv2.bn.running_mean
model.model.6.m.1.cv2.bn.running_var
model.model.6.m.2.cv1.bn.running_mean
model.model.6.m.2.cv1.bn.running_var
model.model.6.m.2.cv2.bn.running_mean
model.model.6.m.2.cv2.bn.running_var
model.model.6.m.3.cv1.bn.running_mean
model.model.6.m.3.cv1.bn.running_var
model.model.6.m.3.cv2.bn.running_mean
model.model.6.m.3.cv2.bn.running_var
model.model.6.m.4.cv1.bn.running_mean
model.model.6.m.4.cv1.bn.running_var
model.model.6.m.4.cv2.bn.running_mean
model.model.6.m.4.cv2.bn.running_var
model.model.6.m.5.cv1.bn.running_mean
model.model.6.m.5.cv1.bn.running_var
model.model.6.m.5.cv2.bn.running_mean
model.model.6.m.5.cv2.bn.running_var
model.model.7.bn.running_mean
model.model.7.bn.running_var
model.model.8.cv1.bn.running_mean
model.model.8.cv1.bn.running_var
model.model.8.cv2.bn.running_mean
model.model.8.cv2.bn.running_var
model.model.8.m.0.cv1.bn.running_mean
model.model.8.m.0.cv1.bn.running_var
model.model.8.m.0.cv2.bn.running_mean
model.model.8.m.0.cv2.bn.running_var
model.model.8.m.1.cv1.bn.running_mean
model.model.8.m.1.cv1.bn.running_var
model.model.8.m.1.cv2.bn.running_mean
model.model.8.m.1.cv2.bn.running_var
model.model.8.m.2.cv1.bn.running_mean
model.model.8.m.2.cv1.bn.running_var
model.model.8.m.2.cv2.bn.running_mean
model.model.8.m.2.cv2.bn.running_var
model.model.9.cv1.bn.running_mean
model.model.9.cv1.bn.running_var
model.model.9.cv2.bn.running_mean
model.model.9.cv2.bn.running_var
model.model.12.cv1.bn.running_mean
model.model.12.cv1.bn.running_var
model.model.12.cv2.bn.running_mean
model.model.12.cv2.bn.running_var
model.model.12.m.0.cv1.bn.running_mean
model.model.12.m.0.cv1.bn.running_var
model.model.12.m.0.cv2.bn.running_mean
model.model.12.m.0.cv2.bn.running_var
model.model.12.m.1.cv1.bn.running_mean
model.model.12.m.1.cv1.bn.running_var
model.model.12.m.1.cv2.bn.running_mean
model.model.12.m.1.cv2.bn.running_var
model.model.12.m.2.cv1.bn.running_mean
model.model.12.m.2.cv1.bn.running_var
model.model.12.m.2.cv2.bn.running_mean
model.model.12.m.2.cv2.bn.running_var
model.model.15.cv1.bn.running_mean
model.model.15.cv1.bn.running_var
model.model.15.cv2.bn.running_mean
model.model.15.cv2.bn.running_var
model.model.15.m.0.cv1.bn.running_mean
model.model.15.m.0.cv1.bn.running_var
model.model.15.m.0.cv2.bn.running_mean
model.model.15.m.0.cv2.bn.running_var
model.model.15.m.1.cv1.bn.running_mean
model.model.15.m.1.cv1.bn.running_var
model.model.15.m.1.cv2.bn.running_mean
model.model.15.m.1.cv2.bn.running_var
model.model.15.m.2.cv1.bn.running_mean
model.model.15.m.2.cv1.bn.running_var
model.model.15.m.2.cv2.bn.running_mean
model.model.15.m.2.cv2.bn.running_var
model.model.16.bn.running_mean
model.model.16.bn.running_var
model.model.18.cv1.bn.running_mean
model.model.18.cv1.bn.running_var
model.model.18.cv2.bn.running_mean
model.model.18.cv2.bn.running_var
model.model.18.m.0.cv1.bn.running_mean
model.model.18.m.0.cv1.bn.running_var
model.model.18.m.0.cv2.bn.running_mean
model.model.18.m.0.cv2.bn.running_var
model.model.18.m.1.cv1.bn.running_mean
model.model.18.m.1.cv1.bn.running_var
model.model.18.m.1.cv2.bn.running_mean
model.model.18.m.1.cv2.bn.running_var
model.model.18.m.2.cv1.bn.running_mean
model.model.18.m.2.cv1.bn.running_var
model.model.18.m.2.cv2.bn.running_mean
model.model.18.m.2.cv2.bn.running_var
model.model.19.bn.running_mean
model.model.19.bn.running_var
model.model.21.cv1.bn.running_mean
model.model.21.cv1.bn.running_var
model.model.21.cv2.bn.running_mean
model.model.21.cv2.bn.running_var
model.model.21.m.0.cv1.bn.running_mean
model.model.21.m.0.cv1.bn.running_var
model.model.21.m.0.cv2.bn.running_mean
model.model.21.m.0.cv2.bn.running_var
model.model.21.m.1.cv1.bn.running_mean
model.model.21.m.1.cv1.bn.running_var
model.model.21.m.1.cv2.bn.running_mean
model.model.21.m.1.cv2.bn.running_var
model.model.21.m.2.cv1.bn.running_mean
model.model.21.m.2.cv1.bn.running_var
model.model.21.m.2.cv2.bn.running_mean
model.model.21.m.2.cv2.bn.running_var
```

**Comment by Burhan-Q** [comment]
@IGlace glad that helped! I'm not 100% certain, but after a little testing and research, I suspect that this might be related to #8753  where the `BatchNorm` modules are not being set to `eval()` mode. Perhaps it would be worth while to check out that pull request (PR) to see if this resolves your concern?

**Comment by IGlace** [comment]
@Burhan-Q thank you very much that worked and solved the problem and i got good results from it, i think you guys should go ahead and push that pull solution to main branch, it's should have been implemented by default honestly after all there is no meaning to freeze layers but keep some of there parameters changing.

Anyway thanks a lot for you help, i have just one more question that i don't know if i should open a new issue for it or just mentioning it here, it's just since i'm manipulating the model weights here from head to another head, from my understanding about the model now, layers from 0-21 are responsible on extracting features of the image in 3 levels, and the final layer 22 is responsible for bounding boxes/classes/segmentation inference.

Now my model will trained only on persons, what i wanna ask is it possible somehow to get those features extracted from the 3 C2F layers [15,18,21] and use them to extract the appearance of each bounding boxes detected in the detection head, and if possible extract only the appearance of the segmented section of the bounding box.

So is this even possible ? or is it already a feature implemented that i don't know ? or it is under developement ?

**Comment by Burhan-Q** [comment]
@IGlace I'm glad it worked! I literally learned about this while trying to help you as it was all new to me 😆! My guess is that the PR will get merged, it is waiting for final review at this point. I've done everything I can to help ensure it has a good chance of being accepted, but the final decision is not up to me. 

As for your question, I'm not entirely clear on what you're asking if I'm honest. Despite this, what I can share with you is that you can use `embeddings = model.predict("path/to/image.png", embed=[15,18,21])` to return the intermediate results from those layers (they will be concatenated and flattened). If you want to get the model outputs from each layer, it'll be a bit more work tho and it's not obvious to me what can be done with these after the fact. The key part of getting the model intermediate layer outputs is going to be that you must format images to be scaled and padded so all dimensions are divisible by 32.

Here's some code on how to collect the intermediate layer outputs:

```py
import cv2 as cv
import numpy as np
import torch

from ultralytics import YOLO

model = YOLO("yolov8n.pt")

im = "ultralytics/assets/bus.jpg"
img = cv.imread(im)
img = img[:640,:640,:] # cropped to size, but should scale and pad to keep all data

img = np.ascontiguousarray(np.transpose(img, (2, 0, 1)))
img = (torch.from_numpy(img).float() / 255.0).unsqueeze(0)

keep = [15, 18, 21] # outputs to keep
save = [6, 4, 12, 9, *keep] # skip connections
y, layer_res = [], []

for m in model.model.model:
    if m.f != -1:
        img = y[m.f] if isinstance(m.f, int) else [img if j == -1 else y[j] for j in m.f]

    # I moved to CUDA as it seemed to help, but might not be required
    _ = m.cuda()
    img = m(img.cuda() if not isinstance(img, list) else [(i.cuda() if i is not None else i) for i in img])

    y.append(img if m.i in save else None)
    _ = layer_res.append(img) if m.i in keep else None

[res.shape for res in layer_res] # view shapes of output data of layers 15, 18, & 21

>>> [torch.Size([1, 64, 80, 80]), torch.Size([1, 128, 40, 40]), torch.Size([1, 256, 20, 20])]
```

From there I'm not sure what you'll want or need to do next (or if that is what you're looking for specifically). Hopefully this is helpful/useful for getting you what you're looking for 🚀

**Comment by IGlace** [comment]
I think it's better to create a new issue for this to discuss it and also for anyone looking for it in the future, i will close this since it's solved.

**Comment by IGlace** [comment]
@Burhan-Q here is the link of the new issue, were we can discuss this more there #9198
